### ISSUE TITLE
Install failure, any suggestion?

### ISSUE THREAD

**Issue by ShixiangWang** [smithlabcode/methpipe#210]
Dear developers,

I encounter errors and don't know how to fix, could you provide some suggestions?

```sh
(wes) [shenao@ctl01 build]$ ../configure
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /usr/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether make supports nested variables... (cached) yes
checking for g++... g++
checking whether the C++ compiler works... yes
checking for C++ compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking for style of include used by make... GNU
checking dependency style of g++... gcc3
checking whether g++ supports C++11 features with -std=c++11... yes
checking for ranlib... ranlib
checking for library containing zlibVersion... no
checking for library containing cblas_dgemm... -lgslcblas
checking for library containing gsl_blas_dgemm... -lgsl
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating config.h
config.status: executing depfiles commands
=== configuring in src/smithlab_cpp (/home/shenao/bioSoft/methpipe-4.1.1/build/src/smithlab_cpp)
configure: running /bin/sh ../../../src/smithlab_cpp/configure --disable-option-checking '--prefix=/usr/local'  --cache-file=/dev/null --srcdir=../../../src/smithlab_cpp
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /usr/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether make supports nested variables... (cached) yes
checking for g++... g++
checking whether the C++ compiler works... yes
checking for C++ compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking for style of include used by make... GNU
checking dependency style of g++... gcc3
checking whether g++ supports C++11 features with -std=c++11... yes
checking for ar... ar
checking the archiver (ar) interface... ar
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking how to print strings... printf
checking for gcc... gcc
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking dependency style of gcc... gcc3
checking for a sed that does not truncate output... /usr/bin/sed
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for fgrep... /usr/bin/grep -F
checking for ld used by gcc... /home/shenao/miniconda3/envs/wes/x86_64-conda-linux-gnu/bin/ld
checking if the linker (/home/shenao/miniconda3/envs/wes/x86_64-conda-linux-gnu/bin/ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm -B
checking the name lister (/usr/bin/nm -B) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 3458764513820540925
checking how to convert x86_64-unknown-linux-gnu file names to x86_64-unknown-linux-gnu format... func_convert_file_noopchecking how to convert x86_64-unknown-linux-gnu file names to toolchain format... func_convert_file_noop
checking for /home/shenao/miniconda3/envs/wes/x86_64-conda-linux-gnu/bin/ld option to reload object files... -r
checking for objdump... objdump
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for archiver @FILE support... @
checking for strip... strip
checking for ranlib... ranlib
checking command to parse /usr/bin/nm -B output from gcc object... ok
checking for sysroot... no
checking for a working dd... /usr/bin/dd
checking how to truncate binary pipes... /usr/bin/dd bs=4096 count=1
checking for mt... no
checking if : is a manifest tool... no
checking how to run the C preprocessor... gcc -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... no
checking for gcc option to produce PIC... -fPIC -DPIC
checking if gcc PIC flag -fPIC -DPIC works... yes
checking if gcc static flag -static works... yes
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/home/shenao/miniconda3/envs/wes/x86_64-conda-linux-gnu/bin/ld -m elf_x86_64) supports shared libraries... yes
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... no
checking whether to build static libraries... yes
checking how to run the C++ preprocessor... g++ -std=c++11 -E
checking for ld used by g++ -std=c++11... /home/shenao/miniconda3/envs/wes/x86_64-conda-linux-gnu/bin/ld -m elf_x86_64
checking if the linker (/home/shenao/miniconda3/envs/wes/x86_64-conda-linux-gnu/bin/ld -m elf_x86_64) is GNU ld... yes
checking whether the g++ -std=c++11 linker (/home/shenao/miniconda3/envs/wes/x86_64-conda-linux-gnu/bin/ld -m elf_x86_64) supports shared libraries... yes
checking for g++ -std=c++11 option to produce PIC... -fPIC -DPIC
checking if g++ -std=c++11 PIC flag -fPIC -DPIC works... yes
checking if g++ -std=c++11 static flag -static works... yes
checking if g++ -std=c++11 supports -c -o file.o... yes
checking if g++ -std=c++11 supports -c -o file.o... (cached) yes
checking whether the g++ -std=c++11 linker (/home/shenao/miniconda3/envs/wes/x86_64-conda-linux-gnu/bin/ld -m elf_x86_64) supports shared libraries... yes
checking dynamic linker characteristics... (cached) GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking for zlibVersion in -lz... no
checking for cblas_dgemm in -lgslcblas... yes
checking for gsl_blas_dgemm in -lgsl... yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating config.h
config.status: executing depfiles commands
config.status: executing libtool commands
(wes) [shenao@ctl01 build]$ make
make  all-recursive
make[1]: Entering directory `/home/shenao/bioSoft/methpipe-4.1.1/build'
Making all in src/smithlab_cpp
make[2]: Entering directory `/home/shenao/bioSoft/methpipe-4.1.1/build/src/smithlab_cpp'
make  all-am
make[3]: Entering directory `/home/shenao/bioSoft/methpipe-4.1.1/build/src/smithlab_cpp'
  CXX      GenomicRegion.lo
  CXX      MappedRead.lo
  CXX      OptionParser.lo
  CXX      QualityScore.lo
  CXX      bisulfite_utils.lo
  CXX      chromosome_utils.lo
  CXX      sim_utils.lo
  CXX      smithlab_os.lo
  CXX      smithlab_utils.lo
  CXX      zlib_wrapper.lo
  CXX      dna_four_bit.lo
  CXXLD    libsmithlab_cpp.la
make[3]: Leaving directory `/home/shenao/bioSoft/methpipe-4.1.1/build/src/smithlab_cpp'
make[2]: Leaving directory `/home/shenao/bioSoft/methpipe-4.1.1/build/src/smithlab_cpp'
make[2]: Entering directory `/home/shenao/bioSoft/methpipe-4.1.1/build'
  CXX      src/common/BetaBin.o
  CXX      src/common/Distro.o
  CXX      src/common/EmissionDistribution.o
  CXX      src/common/Epiread.o
  CXX      src/common/EpireadStats.o
  CXX      src/common/LevelsCounter.o
  CXX      src/common/MethpipeSite.o
  CXX      src/common/Smoothing.o
  CXX      src/common/ThreeStateHMM.o
  CXX      src/common/TwoStateHMM.o
  CXX      src/common/TwoStateHMM_PMD.o
  CXX      src/common/bsutils.o
  CXX      src/common/numerical_utils.o
  AR       libmethpipe.a
  CXX      src/smithlab_cpp/GenomicRegion.o
  CXX      src/smithlab_cpp/MappedRead.o
  CXX      src/smithlab_cpp/OptionParser.o
  CXX      src/smithlab_cpp/QualityScore.o
  CXX      src/smithlab_cpp/bisulfite_utils.o
  CXX      src/smithlab_cpp/chromosome_utils.o
  CXX      src/smithlab_cpp/sim_utils.o
  CXX      src/smithlab_cpp/smithlab_os.o
  CXX      src/smithlab_cpp/smithlab_utils.o
  CXX      src/smithlab_cpp/zlib_wrapper.o
  AR       libsmithlab_cpp.a
  CXX      src/analysis/pmd.o
  CXXLD    pmd
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld:
 src/analysis/pmd.o: in function `check_if_array_data(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
pmd.cpp:(.text+0x1b8b): undefined reference to `gzopen'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x1bac): undefined reference to `gzbuffer'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x1c15): undefined reference to `gzeof'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x207f): undefined reference to `gzclose_r'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: src/analysis/pmd.o: in function `binsize_selection(bool const&, unsigned long&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, double&, double&)':
pmd.cpp:(.text+0x22bf): undefined reference to `gzopen'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x22d8): undefined reference to `gzbuffer'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x2455): undefined reference to `gzeof'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x24b0): undefined reference to `gzeof'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x278e): undefined reference to `gzclose_r'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: src/analysis/pmd.o: in function `optimize_boundaries(unsigned long, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::vector<GenomicRegion, std::allocator<GenomicRegion> >&, std::vector<bool, std::allocator<bool> > const&, std::vector<double, std::allocator<double> > const&, std::vector<double, std::allocator<double> > const&, std::vector<double, std::allocator<double> > const&, std::vector<double, std::allocator<double> > const&)':
pmd.cpp:(.text+0x3303): undefined reference to `gzopen'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x331d): undefined reference to `gzbuffer'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x355d): undefined reference to `gzeof'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x35c7): undefined reference to `gzeof'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x3e2f): undefined reference to `gzclose_r'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x4615): undefined reference to `gzopen'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x462f): undefined reference to `gzbuffer'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x48e5): undefined reference to `gzeof'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x494f): undefined reference to `gzeof'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x4e5e): undefined reference to `gzclose_r'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: src/analysis/pmd.o: in function `load_wgbs_data(unsigned long, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<SimpleGenomicRegion, std::allocator<SimpleGenomicRegion> >&, std::vector<std::pair<double, double>, std::allocator<std::pair<double, double> > >&, std::vector<unsigned long, std::allocator<unsigned long> >&)':
pmd.cpp:(.text+0x66de): undefined reference to `gzopen'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x66f7): undefined reference to `gzbuffer'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x685d): undefined reference to `gzeof'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x68bc): undefined reference to `gzeof'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x6a1d): undefined reference to `gzclose_r'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: src/analysis/pmd.o: in function `load_array_data(unsigned long, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<SimpleGenomicRegion, std::allocator<SimpleGenomicRegion> >&, std::vector<std::pair<double, double>, std::allocator<std::pair<double, double> > >&, std::vector<unsigned long, std::allocator<unsigned long> >&)':
pmd.cpp:(.text+0x750a): undefined reference to `gzopen'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x7523): undefined reference to `gzbuffer'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x7675): undefined reference to `gzeof'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x76d4): undefined reference to `gzeof'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: pmd.cpp:(.text+0x7860): undefined reference to `gzclose_r'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: src/analysis/pmd.o: in function `igzfstream::~igzfstream()':
pmd.cpp:(.text._ZN10igzfstreamD2Ev[_ZN10igzfstreamD5Ev]+0x9): undefined reference to `gzclose_r'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld:
 libsmithlab_cpp.a(zlib_wrapper.o): in function `igzfstream::peek()':
zlib_wrapper.cpp:(.text+0x2f): undefined reference to `gzeof'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: zlib_wrapper.cpp:(.text+0x49): undefined reference to `gzgetc'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: zlib_wrapper.cpp:(.text+0x55): undefined reference to `gzeof'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: zlib_wrapper.cpp:(.text+0x66): undefined reference to `gzungetc'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: libsmithlab_cpp.a(zlib_wrapper.o): in function `getline(igzfstream&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)':
zlib_wrapper.cpp:(.text+0xa0): undefined reference to `gzgets'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: libsmithlab_cpp.a(zlib_wrapper.o): in function `operator<<(ogzfstream&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
zlib_wrapper.cpp:(.text+0x27c): undefined reference to `gzputs'
/home/shenao/miniconda3/envs/wes/bin/../lib/gcc/x86_64-conda-linux-gnu/11.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: libsmithlab_cpp.a(zlib_wrapper.o): in function `operator<<(ogzfstream&, char)':
zlib_wrapper.cpp:(.text+0x29d): undefined reference to `gzputc'
collect2: error: ld returned 1 exit status
make[2]: *** [pmd] Error 1
make[2]: Leaving directory `/home/shenao/bioSoft/methpipe-4.1.1/build'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/home/shenao/bioSoft/methpipe-4.1.1/build'
make: *** [all] Error 2
```

**Comment by guilhermesena1** [comment]
Looks like a problem with zlib linkage. Can you specify which OS you are trying to install on, whether you are using a clone/release/etc, and which commands you used to run? thank you!

**Comment by ShixiangWang** [comment]
@guilhermesena1 Thanks for your response.

I tried installing it on Cent OS7, to build the software, I installed the gcc, g++, zlib, htslib and gsl from conda-forge/bioconda. The strange thing is the make process indicates it bypass the zlib check but failed to find objects shown above.

Currently, I build a conda recipe (<https://github.com/bioconda/bioconda-recipes/pull/34846>) to auto-compile the software and it works. After the  acceptance, the methpipe can be installed from bioconda directly.
